free/src/main/scala/cats/free/Free.scala

```diff
@@ -130,12 +130,17 @@ sealed abstract class Free[S[_], A] extends Product with Serializable {
    * Run to completion, mapping the suspension with the given transformation at each step and
    * accumulating into the monad `M`.
    */
-  final def foldMap[M[_]](f: S ~> M)(implicit M: Monad[M]): M[A] =
+  @tailrec
+  final def foldMap[M[_]](f: S ~> M)(implicit M: Monad[M]): M[A] = {
     step match {
       case Pure(a) => M.pure(a)
       case Suspend(s) => f(s)
-      case Gosub(c, g) => M.flatMap(c.foldMap(f))(cc => g(cc).foldMap(f))
+      case Gosub(c, g) => c match {
+        case Suspend(s) => g(f(s)).foldMap(f)
+        case _ => throw new Error("Unexpected operation. The case should have been eliminated by `step`.")
```

Comment:
Need to check where else is step used, could it be moved to this function to avoid the exception throwing or does it add value on its own?